### PR TITLE
[stdlib] Implement Atomics module

### PR DESCRIPTION
**Pull request by lorentey**
This is the implementation that goes with the [Low-Level Atomic Operations][pitch] pitch. For details, see there!

[pitch]: https://gist.github.com/lorentey/cf8703b5974ebe8f85cfb92a6628880d

In this pitch, we are gathering feedback about the idea of adding a limited set of low-level atomic operations to the Standard Library, including native spellings for C++-style acquire-release memory orderings. Our goal is to enable intrepid library authors to start building synchronization constructs directly in Swift.

As a quick taste, we want to make atomics look like this:

```swift
import Atomics
import Dispatch

let counter = UnsafeAtomic<Int>.create(initialValue: 0)

DispatchQueue.concurrentPerform(iterations: 10) { _ in
  for _ in 0 ..< 1_000_000 {
    counter.wrappingIncrement(ordering: .relaxed)
  }
}
print(counter.load(ordering: .relaxed))
counter.destroy()
```

(This is a followup to the atomic experiments in #27229.)

Fixes #51640
rdar://17171396

### CODE REVIEW

**Comment by lorentey** [comment]
@swift-ci test

File: stdlib/public/Atomics/AtomicMemoryOrderings.swift [hunk 1]
```diff
@@ -0,0 +1,287 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
```
Comment (benrimmington) [inline stdlib/public/Atomics/AtomicMemoryOrderings.swift]: Remove `2014 - ` from all new files? (SR-7507)

```suggestion
// Copyright (c) 2020 Apple Inc. and the Swift project authors
```

File: stdlib/public/Atomics/AtomicMemoryOrderings.swift [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+extension AtomicLoadOrdering {
+  // FIXME: Explain these ordering levels in more detail.
+
+  /// Guarantees the atomicity of the specific operation on which it is applied,
+  /// but imposes no ordering constraints on any other reads or writes.
+  @_semantics("constant_evaluable")
+  @_alwaysEmitIntoClient
+  @_transparent // Debug performance
+  public static var relaxed: AtomicLoadOrdering {
+    AtomicLoadOrdering(_rawValue: 0)
```
Comment (benrimmington) [inline stdlib/public/Atomics/AtomicMemoryOrderings.swift]: These properties and the `Equatable.==` functions could use `Self` from [SE-0068](https://github.com/apple/swift-evolution/blob/master/proposals/0068-universal-self.md).

```suggestion
  public static var relaxed: Self {
    Self(_rawValue: 0)
```

File: stdlib/public/Atomics/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,26 @@
+#===--- CMakeLists.txt - Differentiable programming support library ------===#
```
Comment (benrimmington) [inline stdlib/public/Atomics/CMakeLists.txt]: I think this has been copy-pasted from elsewhere. Is the comment necessary?

```suggestion
#===--- CMakeLists.txt - Low-Level Atomic Operations ---------------------===#
```

**Comment by lorentey** [comment]
@swift-ci test

**Comment by lorentey** [comment]
@swift-ci test

**Comment by lorentey** [comment]
@swift-ci test

**Comment by lorentey** [comment]
@swift-ci test Windows platform

File: stdlib/public/Atomics/AtomicIntegers.swift.gyb [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+    return str[:1].lower() + str[1:] if str else ""
+
+  def argLabel(label):
+    return label + ": " if label <> "_" else ""
+
+  integerOperations = [
+    # Swift name,         llvm name,  operator, label,  doc
+    ('WrappingIncrement', 'add',      '&+',     "by",   "wrapping add"),
+    ('WrappingDecrement', 'sub',      '&-',     "by",   "wrapping subtract"),
+    ('BitwiseAnd',        'and',      '&',      "with", "bitwise and"),
```
Comment (xwu) [inline stdlib/public/Atomics/AtomicIntegers.swift.gyb]: Why the change to "bitwise and with," etc.?

Comment (lorentey) [reply stdlib/public/Atomics/AtomicIntegers.swift.gyb]: The label seems important to clarify the argument’s role in the `atomicLoadThenBitwiseAnd(at:with:ordering:)` variant, and it seems appropriate to apply it to this shorter version, too.

Comment (xwu) [reply stdlib/public/Atomics/AtomicIntegers.swift.gyb]: Have you considered simply reversing the arguments?

```
atomicLoadThenWrappingIncrement(by:at:ordering:)
atomicLoadThenBitwiseAnd(_:at:ordering:)
```

(Incidentally, _TSPL_ calls it "overflow add" instead of "wrapping add"; I do think "wrapping" is more self-explanatory here though.)

File: stdlib/public/Atomics/AtomicRepresentable.swift [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+extension AtomicRepresentable {
+  //public typealias AtomicStorage = RawValue.AtomicStorage
+
+  @inlinable
+  public static func atomicStorage(for value: Self) -> RawValue.AtomicStorage {
+    return RawValue.atomicStorage(for: value.rawValue)
+  }
+
+  @inlinable
+  public static func deinitializeAtomicStorage(
```
Comment (belkadan) [inline stdlib/public/Atomics/AtomicRepresentable.swift]: You could avoid this by requiring that AtomicStorage types be trivial.

File: stdlib/public/Atomics/AtomicPointers.swift.gyb [hunk 6]
```diff
@@ -0,0 +1,514 @@
...
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+%{
+  def atomicTypes():
+    return [
+      # Swift                     AtomicStorage              Builtin
+      ("UnsafeMutableRawPointer", "UnsafeMutableRawPointer", "RawPointer"),
+      ("UnsafeMutablePointer",    "UnsafeMutablePointer",    "RawPointer"),
```
Comment (belkadan) [inline stdlib/public/Atomics/AtomicPointers.swift.gyb]: Why only the mutable variants?

File: stdlib/public/Atomics/AtomicProtocols.swift [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+  @inlinable
+  public static func deinitializeAtomicStorage(
+    at address: UnsafeMutablePointer<AtomicStorage>
+  ) {
+    address.deinitialize(count: 1)
+  }
+}
+
+@available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+public protocol AtomicInteger: AtomicProtocol, FixedWidthInteger {
```
Comment (belkadan) [inline stdlib/public/Atomics/AtomicProtocols.swift]: No atomic increment/decrement for pointers? (Could be useful for lockless data structures.)

File: stdlib/public/Atomics/AtomicIntegers.swift.gyb (hunk 4 above)
Comment (belkadan) [reply stdlib/public/Atomics/AtomicIntegers.swift.gyb]: The order matters; it specifies whether the new or old value is returned.

Comment (lorentey) [reply stdlib/public/Atomics/AtomicIntegers.swift.gyb]: Well, the order is encoded in the name; the argument order doesn't really indicate anything. I like Xiaodi's suggestion, but I think I'd prefer to have the `with:` label for bitwise operations even in the usual case where the address is implicit in `self`:

```
foo.loadThenWrappingIncrement(by: 3, ordering: .relaxed)
bar.loadThenBitwiseAnd(with: baz, ordering: .acquiring)

Int.loadThenBitwiseOr(with: 27, at: ptr, ordering: .releasing)
```

File: stdlib/public/Atomics/AtomicPointers.swift.gyb (hunk 6 above)
Comment (lorentey) [reply stdlib/public/Atomics/AtomicPointers.swift.gyb]: What's the point of an immutable atomic value?

Comment (belkadan) [reply stdlib/public/Atomics/AtomicPointers.swift.gyb]: These are the pointee types, not the pointer types, right? You could certainly have an `UnsafeAtomicPointer<UnsafePointer<ConfigurationCStruct>>`, and swap out configurations that way.

File: stdlib/public/Atomics/AtomicIntegers.swift.gyb (hunk 4 above)
Comment (belkadan) [reply stdlib/public/Atomics/AtomicIntegers.swift.gyb]: Oops, I missed that the comment was only about changing argument labels, not the full name. My mistake!

File: stdlib/public/Atomics/AtomicRepresentable.swift (hunk 5 above)
Comment (lorentey) [reply stdlib/public/Atomics/AtomicRepresentable.swift]: This is preparing for an atomic strong reference type, which will need a custom destructor even if the storage is trivial.

Comment (belkadan) [reply stdlib/public/Atomics/AtomicRepresentable.swift]: But you're deinitializing the storage type, not the value type. That won't ever do anything.

File: stdlib/public/Atomics/AtomicProtocols.swift (hunk 7 above)
Comment (lorentey) [reply stdlib/public/Atomics/AtomicProtocols.swift]: Hm -- `AtomicAdditiveArithmetic`?

Comment (belkadan) [reply stdlib/public/Atomics/AtomicProtocols.swift]: AtomicStrideable, I think. That's what I put in my prototype, and it can't be AdditiveArithmetic because the types don't match up.

Comment (lorentey) [reply stdlib/public/Atomics/AtomicProtocols.swift]: Sounds good! So we'll have this tree of protocols:

```
AtomicProtocol → AtomicStrideable → AtomicInteger
                           → NullableAtomic
```

Comment (lorentey) [reply stdlib/public/Atomics/AtomicProtocols.swift]: Except not at all -- the stride needs to be independent of `Self` :-P

Comment (lorentey) [reply stdlib/public/Atomics/AtomicProtocols.swift]: Well, I guess `AtomicInteger` could constrain `Stride == Self`. 🤔

Comment (belkadan) [reply stdlib/public/Atomics/AtomicProtocols.swift]: :-/ In my API prototype I had the backing storage of pointers be Int for this reason. The strideable-ness is also still important because the unit stride of, e.g. `UnsafePointer<Int32>` is 4, not 1.

File: stdlib/public/Atomics/AtomicRepresentable.swift (hunk 5 above)
Comment (lorentey) [reply stdlib/public/Atomics/AtomicRepresentable.swift]: This is the teardown of the logical atomic value -- if `Self` is modeling a strong reference, then it needs to have a hook to release the value.

Comment (belkadan) [reply stdlib/public/Atomics/AtomicRepresentable.swift]: Ah, okay. In that case, the implementation is wrong, since it's calling `deinitialize` on an `UnsafeMutablePointer<AtomicStorage>` and not `UnsafeMutablePointer<Self>` (or `UnsafeMutablePointer<RawValue>`).

File: stdlib/public/Atomics/AtomicProtocols.swift (hunk 7 above)
Comment (lorentey) [reply stdlib/public/Atomics/AtomicProtocols.swift]: Hm -- I keep flip-flopping on backing pointers/`Unmanaged` with `Int`, but implementing arithmetic doesn't seem like a huge issue either way.

We need to be able to model integer types narrower (and probably wider) than `Int` though.

Comment (belkadan) [reply stdlib/public/Atomics/AtomicProtocols.swift]: \*shrug\* Requiring that `Stride` be the same as `Self` seems reasonable for AtomicInteger, but you could also just make it a default implementation for the map-Stride-into-AtomicStorage operation on AtomicStrideable (if you go with my route).

Comment (lorentey) [reply stdlib/public/Atomics/AtomicProtocols.swift]: Let's see where the implementation takes us. 

Hm, are we to define pointer arithmetic for `Optional<UnsafePointer>`, too?

Comment (belkadan) [reply stdlib/public/Atomics/AtomicProtocols.swift]: Here was my unhappy best attempt:

```
// Optional does not (and should not) conform to Strideable...
// ...but it seems useful to be able to move pointers around atomically...
// ...but it's not type-safe when the value is nil.
//
// Actually, more specifically: loadThenIncrement can trap on nil, but incrementThenLoad can't.
//
// So we could do something like this (using Alejandro Alonso's work on parameterized extensions, https://forums.swift.org/t/parameterized-extensions/25563).

#if compiler(>=6)
extension <Wrapped> UnsafeAtomicPointer where
    Value == Optional<Wrapped>,
    Wrapped: AtomicStrideable,
    Wrapped: AtomicOptionalRepresentable {
  // The two meanings of "wrapped" here is unfortunate.
  func loadUnwrappedThenWrappingIncrement(
    by operand: Wrapped.Stride,
    ordering: AtomicUpdateOrdering
  ) -> Wrapped {
    let rawStride = Wrapped.atomicStorageRepresentation(of: operand)
    let rawResult = address.withMemoryRebound(to: Value.AtomicStorage.self, capacity: 1) {
      Value.AtomicStorage.atomicLoadThenWrappingIncrement(from: $0, by: rawStride, ordering: ordering)
    }
    // Note that we're still checking if the loaded value is nil.
    return unsafeBitCast(result, to: Optional<Wrapped>.self)!
  }
}
#endif
```

File: stdlib/public/Atomics/AtomicRepresentable.swift (hunk 5 above)
Comment (lorentey) [reply stdlib/public/Atomics/AtomicRepresentable.swift]: We don't have an atomic strong reference yet*. The existing atomic types are all trivial, so they could just as well have an empty implementation for `deinitializeAtomicStorage`.

* `UnsafeAtomicLazyReference` doesn't conform to `AtomicProtocol`, and (as of this moment) it has non-trivial storage.

Comment (belkadan) [reply stdlib/public/Atomics/AtomicRepresentable.swift]: Ohhh, it's future-proofing. I get it now. Thanks for explaining it, sorry it took so long for me to understand!

File: stdlib/public/Atomics/AtomicProtocols.swift (hunk 7 above)
Comment (lorentey) [reply stdlib/public/Atomics/AtomicProtocols.swift]: I think we mustn't allow the results of nil pointer arithmetic to get committed into memory -- it would allow other threads to go astray before the trap happens. (So the `loadThen` and the `ThenLoad` forms must always have the same observable effect.)

Parameterized extensions can be replaced by more protocols; but is this really something that we need to model?

Comment (belkadan) [reply stdlib/public/Atomics/AtomicProtocols.swift]: Oof, I didn't think about other threads doing work before the trap. Maybe there's no safe way to do this. It's really too bad, though, because a compare-and-swap-based update will be needlessly slower.

(We could also say "it's undefined behavior to increment a nil pointer" but even for an unsafe API that's pretty bad.)

Comment (lorentey) [reply stdlib/public/Atomics/AtomicProtocols.swift]: Wait -- what's the envisioned use case for atomic pointer arithmetic again?

The pointer will necessarily point into some bounded buffer, and a direct atomic +1 won't be able to do any bounds checking before incrementing the value -- so if there is a high enough number of competing threads, they could cause the pointer to escape its buffer.

Comment (belkadan) [reply stdlib/public/Atomics/AtomicProtocols.swift]: You could do a fast atomic increment and then check for validity afterward, but yeah, maybe I'm just hypothesizing things people don't actually do. Maybe someone else on the thread can be called in to check?

Comment (lorentey) [reply stdlib/public/Atomics/AtomicProtocols.swift]: That runs into the same problem as null pointer arithmetic above -- we must not allow invalid state to get atomically committed into memory. It sounds like pointer arithmetic will need to be always done with explicit compare/exchange loops. (Unless CPU vendors vastly expand their atomic instruction sets...)

Comment (belkadan) [reply stdlib/public/Atomics/AtomicProtocols.swift]: Yeah, oh well. Maybe that means AtomicStrideable's not worth it. (On the other hand, if it turns out to be useful, we don't yet have a way to add it in a later release…)

Comment (lorentey) [reply stdlib/public/Atomics/AtomicProtocols.swift]: I'm pretty sure we'll be able to eventually add it if CPUs get bounded atomic increments -- it's "just" a matter of adding a new protocol or two.

**Comment by lorentey** [comment]
@swift-ci test

File: stdlib/public/Atomics/AtomicIntegerConformances.swift.gyb [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+  @_semantics("has_constant_evaluable_arguments")
+  @_transparent @_alwaysEmitIntoClient
+  public static func atomicCompareExchange(
+    expected: ${swiftType},
+    desired: ${swiftType},
+    at address: UnsafeMutablePointer<${swiftType}>,
+    ordering: AtomicUpdateOrdering,
+    failureOrdering: AtomicLoadOrdering
+  ) -> (exchanged: Bool, original: ${swiftType}) {
+    switch (ordering, failureOrdering) {
```
Comment (glessard) [inline stdlib/public/Atomics/AtomicIntegerConformances.swift.gyb:228]: Is this one also a strong CAS?

Comment (lorentey) [reply stdlib/public/Atomics/AtomicIntegerConformances.swift.gyb:228]: Ah, yes, it is! I'll need to do another pass on updating the docs.

File: stdlib/public/Atomics/AtomicPointers.swift.gyb (hunk 6 above)
Comment (lorentey) [reply stdlib/public/Atomics/AtomicPointers.swift.gyb]: `UnsafeRawPointer` and `UnsafePointer` both conform to `NullableAtomic` now.

File: stdlib/public/Atomics/AtomicIntegerConformances.swift.gyb (hunk 8 above)
Comment (lorentey) [reply stdlib/public/Atomics/AtomicIntegerConformances.swift.gyb:172]: Done; I also added a separate weak variant (two orderings only, for now).

File: stdlib/public/Atomics/UnsafeAtomicLazyReference.swift [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+@available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+extension UnsafeAtomicLazyReference {
+  @inlinable
+  public static func create() -> Self {
+    let ptr = UnsafeMutablePointer<AtomicStorage>.allocate(capacity: 1)
+    ptr.initialize(to: nil)
+    return Self(at: ptr)
+  }
+
+  public func destroy() {
```
Comment (glessard) [inline stdlib/public/Atomics/UnsafeAtomicLazyReference.swift:41]: Should you release a non-nil unmanaged here?

File: stdlib/public/Atomics/AtomicIntegerConformances.swift.gyb [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+% for (swiftSuccess, llvmSuccess, _) in rmwOrderings():
+%   for (swiftFailure, llvmFailure) in loadOrderings():
+    case (.${swiftSuccess}, .${swiftFailure}):
+      let (oldValue, won) = Builtin.cmpxchg_${llvmOrders(llvmSuccess, llvmFailure)}_${builtinType}(
+        pointer._rawValue, expected.${getter}, desired.${getter})
+      return (Bool(_builtinBooleanLiteral: won), ${swiftType}(oldValue))
+%   end
+% end
+    default:
+      let (oldValue, won) = Builtin.cmpxchg_seqcst_seqcst_${builtinType}(
```
Comment (gribozavr) [inline stdlib/public/Atomics/AtomicIntegerConformances.swift.gyb:182]: Shouldn't the `default:` just contain some kind of unreachable marker?

File: stdlib/public/Atomics/AtomicIntegerConformances.swift.gyb [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+% for (swiftSuccess, llvmSuccess, _) in rmwOrderings():
+%   for (swiftFailure, llvmFailure) in loadOrderings():
+    case (.${swiftSuccess}, .${swiftFailure}):
+      let (oldValue, won) = Builtin.cmpxchg_${llvmOrders(llvmSuccess, llvmFailure)}_weak_${builtinType}(
+        pointer._rawValue, expected.${getter}, desired.${getter})
+      return (Bool(_builtinBooleanLiteral: won), ${swiftType}(oldValue))
+%   end
+% end
+    default:
+      let (oldValue, won) = Builtin.cmpxchg_seqcst_seqcst_weak_${builtinType}(
```
Comment (gribozavr) [inline stdlib/public/Atomics/AtomicIntegerConformances.swift.gyb:207]: Shouldn't the `default:` just contain some kind of unreachable marker?

File: stdlib/public/Atomics/UnsafeAtomicLazyReference.swift (hunk 9 above)
Comment (lorentey) [reply stdlib/public/Atomics/UnsafeAtomicLazyReference.swift:41]: No -- unmanaged references aren't necessarily owning (a.k.a +1). If they are, they need to manually released before deallocation.

File: stdlib/public/Atomics/AtomicIntegerConformances.swift.gyb (hunk 10 above)
Comment (lorentey) [reply stdlib/public/Atomics/AtomicIntegerConformances.swift.gyb:182]: Good point. This is a remnant of a CRISP pattern from a previous design iteration.

Comment (lorentey) [reply stdlib/public/Atomics/AtomicIntegerConformances.swift.gyb:182]: Good point. (This is a remnant of a CRISP from a previous design iteration.)

File: stdlib/public/Atomics/UnsafeAtomicLazyReference.swift (hunk 9 above)
Comment (glessard) [reply stdlib/public/Atomics/UnsafeAtomicLazyReference.swift:41]: Now I see how it happens. Don't try this at home!

File: test/Atomics/AtomicsFolding.swift.gyb [hunk 12]
```diff
@@ -0,0 +1,562 @@
...
+    ordering: .sequentiallyConsistent)
+}
+// SECTIONING-LABEL: } // end sil function '$s14AtomicsFolding
+
+% end
+
+% for (operation, modifier) in [("compareExchange", ""), ("weakCompareExchange", "_weak")]:
+// -----------------------------------------------------------------------------
+// FIXME: There are 15 valid combinations of success/failure orderings;
+// here we're only testing 6 of them.
```
Comment (varungandhi-apple) [inline test/Atomics/AtomicsFolding.swift.gyb:336]: Is this intended to be fixed before merging? I see there's a

```
FIXME: Swift should provide intrinsics for arbitrary ordering pairs
```

at the beginning -- I'm not entirely sure if the two are related. If they are, please add a note or something indicating the relation between the two. For example, you could have something like `FIXME: [limited-ordering-pairs]` in both places so one can easily connect the two.

Comment (lorentey) [reply test/Atomics/AtomicsFolding.swift.gyb:336]: > Is this intended to be fixed before merging?

Maybe -- the obvious way to add the missing test cases would be to stamp out all the combinations with gyb, but this test is already unreadably meta. It's not clear that adding full coverage is necessary here; I can't imagine the compiler would decide to arbitrarily refuse to optimize just one of the argument combinations. But what do I know... ¯\_(ツ)_/¯

> I'm not entirely sure if the two are related.

They aren't! We can stop "maxing" success ordering to the failure ordering when we switch to an llvm version that supports independent success/failure orderings. (I believe this might be on the way, since C++ looks set to lift the corresponding constraint soon.)

Comment (varungandhi-apple) [reply test/Atomics/AtomicsFolding.swift.gyb:336]: > Maybe -- the obvious way to add the missing test cases would be to stamp out all the combinations with gyb, but this test is already unreadably meta. It's not clear that adding full coverage is necessary here; I can't imagine the compiler would decide to arbitrarily refuse to optimize just one of the argument combinations. But what do I know... ¯\\_(ツ)_/¯

Ok... there's two arguments here. (1) It is better to be safe than sorry (so add the missing combinations) and (2) The value add of those combinations is negligible (so not add the missing combinations). Personally I would lean towards (1), but I can understand if you think (2) is more important here.

In any case, leaving the FIXME kinda' makes it seems like there is something to be fixed, but that doesn't seem to be the intention. So my thought here is:

1. We get rid of the FIXME and adding the missing combinations.
2. OR We add a reason for why the combinations are missing, and remove the FIXME, just have it be a normal comment (indicating that it is deliberate decision and not for lack of time).

Comment (lorentey) [reply test/Atomics/AtomicsFolding.swift.gyb]: FIXMEs do not come with an expectation that we need (or even want) them to be fixed -- they merely note an aspect that we could perhaps revisit, if we had no time/resource/ABI constraints.

The test serves it intended purpose as it is! Adding test coverage for all ordering combinations is non-essential work, and we can easily do it (or not) at any point. The FIXME does its job by calling out the missing combos.

Short term, it would probably be more important to work on adding different tests, especially tests that actually exercise atomicity and memory orderings.

**Comment by freak4pc** [comment]
This looks great, thanks for working on it! Looking forward to seeing this added, and help avoid many hacks non-official implementations. 

I do wonder though if there are any plans for Atomicity in a type-safe level, as opposed to this UnsafeAtomic implementation. Not necessarily in this implementation, but generally speaking ? 

Thanks !

**Comment by lorentey** [comment]
Yes -- the current implementation was specifically designed to allow easy transition to the eventual memory-safe constructs later. The preliminary proposal text [has a section explaining what's missing](https://gist.github.com/lorentey/cf8703b5974ebe8f85cfb92a6628880d#memory-safe-atomic-constructs).

File: stdlib/public/Atomics/UnsafeAtomicLazyReference.swift (hunk 9 above)
Comment (lorentey) [reply stdlib/public/Atomics/UnsafeAtomicLazyReference.swift]: The latest revision switches atomic lazy references to use trivial storage representation, so it is now critical to call `dispose` before deinitializing storage.

**Comment by lorentey** [comment]
@swift-ci test

File: stdlib/public/Atomics/UnsafeAtomicLazyReference.swift [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+//===----------------------------------------------------------------------===//
+
+import Swift
+
+/// An atomic optional strong reference that can be set (initialized) exactly
+/// once, but read many times.
+@available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+@frozen
+public struct UnsafeAtomicLazyReference<Instance: AnyObject> {
+  public typealias Value = Instance?
```
Comment (benrimmington) [inline stdlib/public/Atomics/UnsafeAtomicLazyReference.swift:20]: It might be clearer to remove this typealias, and use `Instance?` directly.

(`Instance` and `Instance?` are already used in the store/load methods.)

File: stdlib/public/Atomics/UnsafeAtomic.swift.gyb [hunk 14]
```diff
@@ -0,0 +1,376 @@
...
+  /// - Returns: A tuple `(exchanged, original)`, where `exchanged` is true if
+  ///   the exchange was successful, and `original` is the original value.
+  @_semantics("atomics.requires_constant_orderings")
+  @_transparent @_alwaysEmitIntoClient
+  public func compareExchange(
+    expected: Value,
+    desired: __owned Value,
+    ordering: AtomicUpdateOrdering,
+    failureOrdering: AtomicLoadOrdering
+  ) -> (exchanged: Bool, original: Value) {
```
Comment (benrimmington) [inline stdlib/public/Atomics/UnsafeAtomic.swift.gyb:225]: In these APIs, I find some of the names confusing.

* `expected` and `original` ➡️ `oldValue`
* `desired` ➡️ `newValue`
* `exchanged` ➡️ `success`
* `ordering` ➡️ `successOrdering` (iff the API has `failureOrdering`)
* possibly an `inout` parameter?

```suggestion
  public func compareExchange(
    oldValue: inout Value,
    newValue: __owned Value,
    successOrdering: AtomicUpdateOrdering,
    failureOrdering: AtomicLoadOrdering
  ) -> Bool {
```

File: stdlib/public/Atomics/AtomicOptional.swift [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+public protocol _PrimitiveAtomicOptional: _PrimitiveAtomic {
+  var _isNil: Bool { get }
+  static var _nilValue: Self { get }
+}
+
+@available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+extension Optional: AtomicProtocol
+where Wrapped: AtomicProtocol, Wrapped._AtomicStorage: _PrimitiveAtomicOptional
+{
+  public typealias _AtomicStorage = Wrapped._AtomicStorage
```
[the remark this answers is not in the document]
Comment (lorentey) [reply stdlib/public/Atomics/AtomicOptional.swift:25]: Yes, but that doesn't look like a particularly great idea to me. I prefer the explicit un-cleverness of a dedicated protocol here.

There is also this:

> `nil` has a specific meaning in Swift—the absence of a value. Only the `Optional` type conforms to `ExpressibleByNilLiteral`. `ExpressibleByNilLiteral` conformance for types that use nil for other purposes is discouraged.

File: stdlib/public/Atomics/UnsafeAtomic.swift.gyb [hunk 16]
```diff
@@ -0,0 +1,376 @@
...
+    public init(_ value: __owned Value) {
+      self._value = Value._prepareAtomicStorage(for: value)
+    }
+
+    @inlinable @inline(__always)
+    @discardableResult
+    public mutating func dispose() -> Value {
+      return Value._disposeAtomicStorage(&_value)
+    }
+  }
```
[the remark this answers is not in the document]
Comment (lorentey) [reply stdlib/public/Atomics/UnsafeAtomic.swift.gyb:61]: Again, this is technically possible, but I prefer to have a dedicated, public wrapper so that particular choices for `_AtomicStorage` won't get hardwired into (ostensibly) portable codebases.

`UnsafeAtomic.Storage` has the additional benefit that we can attach API to it (`init(_:)`, `dispose()`) that would be inappropriate on `_AtomicStorage`, esp. in the case where atomic types use `Self` as their atomic storage.

File: stdlib/public/Atomics/AtomicOptional.swift (hunk 15 above)
[the remark this answers is not in the document]
Comment (benrimmington) [reply stdlib/public/Atomics/AtomicOptional.swift:25]: > There is also this:
> 
> > `nil` has a specific meaning in Swift—the absence of a value. Only the `Optional` type conforms to `ExpressibleByNilLiteral`.

And the [`_OptionalNilComparisonType`](https://github.com/apple/swift/blob/b3b7aa8ecfa3cdf3f52544597f3ce0d7ebe03a77/stdlib/public/core/Optional.swift#L398-L406)!

File: stdlib/public/Atomics/UnsafeAtomicLazyReference.swift (hunk 13 above)
Comment (lorentey) [reply stdlib/public/Atomics/UnsafeAtomicLazyReference.swift:20]: `Value` is part of the new unsafe API pattern for high-performance synchronization primitives. This is pure 100% ceremony, but it helps establishing / highlighting patterns for such types.

File: stdlib/public/Atomics/AtomicOptional.swift (hunk 15 above)
[the remark this answers is not in the document]
Comment (belkadan) [reply stdlib/public/Atomics/AtomicOptional.swift:25]: I do think it'd be fair to say `_PrimitiveAtomicOptional` *inherits* from `ExpressibleByNilLiteral` though. That would still simplify the implementation (at the cost of not preventing people from conforming to it…)

File: stdlib/public/Atomics/UnsafeAtomic.swift.gyb (hunk 14 above)
Comment (lorentey) [reply stdlib/public/Atomics/UnsafeAtomic.swift.gyb:225]: I like `successOrdering` -- I'm pushing a commit that switches to that.

Having an `inout` parameter was deliberately rejected, based on negative experience with the corresponding C++ construct.

I don't see how `expected`/`desired` would be confusing, and they match preexisting terminology. But this may be a topic for the evolution thread.

**Comment by lorentey** [comment]
@swift-ci test

File: stdlib/public/Atomics/AtomicOptional.swift (hunk 15 above)
[the remark this answers is not in the document]
Comment (lorentey) [reply stdlib/public/Atomics/AtomicOptional.swift:25]: The storage types aren't currently `ExpressibleByNilLiteral`, though, and I _think_ I'd like to keep them that way. 

I am ambivalent towards adding `Equatable` conformance to atomic storage. (For the same reason I keep fiddling with the `FixedWidthInteger` conformance on `_PrimitiveAtomicInteger`. I dislike adding too much structure to the storage types -- these should be treated as black boxes.)

**Comment by lorentey** [comment]
@swift-ci test

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Comment by lorentey** [comment]
Closing -- Atomics are published as a standalone package for now: https://github.com/apple/swift-atomics
